Feature blog for Release 1.20: Kubernetes Volume Snapshot Moves to GA (#25445)
kubernetes/website: content/en/blog/_posts/2020-12-10-Kubernetes Volume Snapshot Moves to GA.md

---
layout: blog
title: 'Kubernetes 1.20: Kubernetes Volume Snapshot Moves to GA'
date: 2020-12-10

> slug: kubernetes-release-1.20-Kubernetes-Volume-Snapshot-Moves-to-GA

Reviewer: How about:
```suggestion
slug: kubernetes-1.20-volume-snapshot-moves-to-ga
```
?

Author: This makes total sense! I thought it was repetitive too :)

The Kubernetes Storage SIG identified snapshot operations as critical functionality for many stateful workloads. For example, a database administrator may want to snapshot a database’s volumes before starting a database operation.

By providing a standard way to trigger volume snapshot operations in Kubernetes, this feature allows Kubernetes users to incorporate snapshot operations in a portable manner on any Kubernetes environment regardless of the underlying storage.

Additionally, these Kubernetes snapshot primitives act as basic building blocks that unlock the ability to develop advanced enterprise-grade storage administration features for Kubernetes,  including application or cluster level backup solutions.

## What’s new since beta?

With the promotion of Volume Snapshot to GA, the feature is enabled by default on standard Kubernetes deployments and cannot be turned off.

Many enhancements have been made to improve the quality of this feature and to make it production-grade.

> - The Volume Snapshot APIs and client library were moved to a separate go module.

Reviewer: nit
```suggestion
- The Volume Snapshot APIs and client library were moved to a separate Go module.
```

Author: Addressed this in the latest commit, PTAL.

> - A snapshot validation webhook has been added to perform necessary validation on volume snapshot objects. More details can be found [here](https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1900-volume-snapshot-validation-webhook).

Reviewer 2: ```suggestion
- A snapshot validation webhook has been added to perform necessary validation on volume snapshot objects. More details can be found in the [Volume Snapshot Validation Webhook Kubernetes Enhancement Proposal.](https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1900-volume-snapshot-validation-webhook).
```
nit: avoids "here".

Author: Added this in the latest commit, PTAL.

> - Along with the validation webhook, the volume snapshot controller will start labeling invalid snapshot objects which already existed. This allows users to identify, remove any invalid objects, and correct their workflows. Once the API is switched to the v1 type, those invalid objects will not be deletable from the system.

Reviewer 2: ```suggestion
- Along with the validation webhook, the volume snapshot controller will start labeling invalid snapshot objects that already existed. This allows users to identify, remove any invalid objects, and correct their workflows. Once the API is switched to the v1 type, those invalid objects will not be deletable from the system.
```
non blocking, just thought it read better 👍

Author: Absolutely agree, it does read better! I have made changes in the latest commit.

> - To provide better insights how the snapshot feature is performing, an initial set of operation metrics has been added to the volume snapshot controller.

Reviewer 2: ```suggestion
- To provide better insights into how the snapshot feature is performing, an initial set of operation metrics has been added to the volume snapshot controller.
```

Author: Definitely an overlook on my part, thank you for flagging this!

> - More e2e tests have been added to test the feature in a real Kubernetes cluster on GCP. Stress tests (based on Google Persistent Disk and hostpath CSI Drivers) have been introduced to test the robustness of the system.

Reviewer: ```suggestion
- There are more end-to-end tests, running on GCP, that validate the feature in a real Kubernetes cluster. Stress tests (based on Google Persistent Disk and `hostPath` CSI Drivers) have been introduced to test the robustness of the system.
```
?

Reviewer 2: +1 to Tim's suggestion, it makes the statement more clear.

Author: Addressed this in the latest commit, PTAL.

Other than introducing tightening validation, there is no difference between the v1beta1 and v1 Kubernetes volume snapshot API. In this release (with Kubernetes 1.20), both v1 and v1beta1 are served while the stored API version is still v1beta1. Future releases will switch the stored version to v1 and gradually remove v1beta1 support.

## Which CSI Drivers support Volume Snapshots?

> Snapshots are only supported for CSI drivers, not for in-tree or Flex. Ensure the deployed CSI driver on your cluster has implemented the snapshot interfaces. For more information, see Container Storage Interface (CSI) for Kubernetes GA.

Reviewer: ```suggestion
Snapshots are only supported for CSI drivers, not for in-tree or FlexVolume drivers. Ensure the deployed CSI driver on your cluster has implemented the snapshot interfaces. For more information, see Container Storage Interface (CSI) for Kubernetes GA.
```
?
(I presume _Container Storage Interface (CSI) for Kubernetes GA_ is also supposed to be a link?)

Reviewer 3: https://kubernetes.io/blog/2019/01/15/container-storage-interface-ga/

Reviewer: ```suggestion
Snapshots are only supported for CSI drivers, not for in-tree or FlexVolume drivers. Ensure the deployed CSI driver on your cluster has implemented the snapshot interfaces. For more information, see [Container Storage Interface (CSI) for Kubernetes GA](https://kubernetes.io/blog/2019/01/15/container-storage-interface-ga/).
```

Author: Addressed this in the latest commit, PTAL.

Currently more than [50 CSI drivers](https://kubernetes-csi.github.io/docs/drivers.html) support the Volume Snapshot feature. The [GCE Persistent Disk CSI Driver](https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver) has gone through the tests for upgrading from volume snapshots beta to GA. GA level support for other CSI drivers should be available soon.

## Who builds products using Volume Snapshots?

> As of the publishing of this blog, following participants from the [Kubernetes Data Protection WG](https://github.com/kubernetes/community/tree/master/wg-data-protection) are building products or have already built products using Kubernetes volume snapshots.

Reviewer 2: ```suggestion
As of the publishing of this blog, the following participants from the [Kubernetes Data Protection Working Group](https://github.com/kubernetes/community/tree/master/wg-data-protection) are building products or have already built products using Kubernetes volume snapshots.
```

- Druva
- [Kasten K10](https://www.kasten.io/)
- Pure Storage (Pure Service Orchestrator)
- Red Hat OpenShift Container Storage
- [TrilioVault for Kubernetes](https://docs.trilio.io/kubernetes/)
- [Velero plugin for CSI](https://github.com/vmware-tanzu/velero-plugin-for-csi)

## How to deploy Volume Snapshots?

Volume Snapshot feature contains the following components:

- [Kubernetes Volume Snapshot CRDs](https://github.com/kubernetes-csi/external-snapshotter/tree/master/client/config/crd)
- [Volume snapshot controller](https://github.com/kubernetes-csi/external-snapshotter/tree/master/pkg/common-controller)
- [Snapshot validation webhook](https://github.com/kubernetes-csi/external-snapshotter/tree/master/pkg/validation-webhook)
- CSI Driver along with [CSI Snapshotter sidecar](https://github.com/kubernetes-csi/external-snapshotter/tree/master/pkg/sidecar-controller)

It is strongly recommended that Kubernetes distributors bundle and deploy the volume snapshot controller, CRDs, and validation webhook as part of their Kubernetes cluster management process (independent of any CSI Driver).

> ### WARNING:

Reviewer: We could use a `{{< warning >}}` shortcode here rather than this heading.

Author: Addressed this in the latest commit, PTAL.

The snapshot validation webhook serves as a critical component to transition smoothly from using v1beta1 to v1 API. Not installing the snapshot validation webhook makes prevention of invalid volume snapshot objects from creation/updating impossible, which in turn will block deletion of invalid volume snapshot objects in coming upgrades.

> If your cluster does not come pre-installed with the correct components, you may manually install them. See details [here](https://github.com/kubernetes-csi/external-snapshotter/blob/master/README.md).

Reviewer: ```suggestion
If your cluster does not come pre-installed with the correct components, you may manually install them. See the [CSI Snapshotter](https://github.com/kubernetes-csi/external-snapshotter#readme) README for details.
```

Author: Addressed this in the latest commit, PTAL.

## How to use Volume Snapshots?

> Assuming all the required components (including CSI driver) have been already deployed and running on your cluster, you can create volume snapshots using the `VolumeSnapshot` API object, or use an existing `VolumeSnapshot` to restore a PVC by specifying the VolumeSnapshot data source on it. For more details, see document [here] https://kubernetes.io/docs/concepts/storage/volume-snapshots/.

Reviewer 2: ```suggestion
Assuming all the required components (including CSI driver) have been already deployed and running on your cluster, you can create volume snapshots using the `VolumeSnapshot` API object, or use an existing `VolumeSnapshot` to restore a PVC by specifying the VolumeSnapshot data source on it. For more details, see the [volume snapshot documentation](/docs/concepts/storage/volume-snapshots/).
```
I think this was intended to be a link^^;;

> ***NOTE***: The Kubernetes Snapshot API does not provide any application consistency guarantees. You have to prepare your application (pause application, freeze filesystem etc.) before taking the snapshot for data consistency either manually or using higher level APIs/controllers.

Reviewer: We could use a `{{< note >}}` shortcode here.

Author: Addressed this in the latest commit, PTAL.

### Dynamically Provision a Volume Snapshot

To dynamically provision a volume snapshot, create a `VolumeSnapshotClass` API object first.

```yaml
apiVersion: snapshot.storage.k8s.io/v1
kind: VolumeSnapshotClass
metadata:
  name: test-snapclass
driver: testdriver.csi.k8s.io
deletionPolicy: Delete
parameters:
  csi.storage.k8s.io/snapshotter-secret-name: mysecret
  csi.storage.k8s.io/snapshotter-secret-namespace: mysecretnamespace
```  

Then create a `VolumeSnapshot` API object from a PVC by specifying the volume snapshot class.

```yaml
apiVersion: snapshot.storage.k8s.io/v1
kind: VolumeSnapshot
metadata:
  name: test-snapshot
  namespace: ns1
spec:
  volumeSnapshotClassName: test-snapclass
  source:
    persistentVolumeClaimName: test-pvc
```

> ### Importing an existing volume snapshot with Kubernetes

Reviewer: Usually we write headings in the page in sentence case. I would either fix all the others, or amend this one to match.

Author: Addressed this in the latest commit, PTAL. I'm unsure if this was what you were looking for. Do let me know if it works.

There are many more people who have helped to move the snapshot feature from beta to GA. We want to thank everyone who has contributed to this effort:
- [Andi Li](@AndiLi99)
- [Ben Swartzlander](@bswartz)
- [Chris Henzie](@chrishenzie)
- [Christian Huffman](@huffmanca)
- [Grant Griffiths](@ggriffiths)
- [Humble Devassy Chirammal](@humblec)
- [Jan Šafránek](@jsafrane)
- [Jiawei Wang](@jiawei0277)
- [Jing Xu](@jingxu97)
- [Jordan Liggitt](@liggitt)
- [Kartik Sharma](@Kartik494)
- [Madhu Rajanna](@Madhu-1)
- [Manohar Reddy](@boddumanohar)
- [Michelle Au](@msau42)
- [Patrick Ohly](@pohly)
- [Prafull Ladha](@prafull01)
- [Prateek Pandey](@prateekpandey14)
- [Raunak Shah](@RaunakShah)
- [Saad Ali](@saadali)
- [Saikat Roychowdhury](@saikat-royc)
- [Tim Hockin](@thockin)
- [Xiangqian Yu](@yuxiangqian)
- [Xing Yang](@xing-yang)
- [Zhu Can](@zhucan)

> For those interested in getting involved with the design and development of CSI or any part of the Kubernetes Storage system, join the [Kubernetes Storage Special Interest Group (SIG)](https://github.com/kubernetes/community/tree/master/sig-storage). We’re rapidly growing and always welcome new contributors.

Reviewer: nit:
```suggestion
For those interested in getting involved with the design and development of CSI or any part of the Kubernetes Storage system, join the [Kubernetes Storage Special Interest Group](https://github.com/kubernetes/community/tree/master/sig-storage)  (SIG). We’re rapidly growing and always welcome new contributors.
```

Author: Addressed this in the latest commit, PTAL.
